HughV2: add drop down in controller to choose server

rmn30 opened the pull request:
This allows selecting 10.0.0.10 for when using a local MQTT server set up with a wifi bridge to ethernet for sonata.
Out of laziness I used the unencrypted port 1883 so that I don't have to set up a custom CA.
Sonata will still use SSL and there is also e2e encryption...

Also add a connect button so you can select the server before attempting to connect.

Sonata firmware tweaks to follow.

rmn30 committed:
HughV2: add drop down in controller to choose server

rmn30 commented:
Unrelated but we should really be using the async API of MQTT3Client as right now it blocks the main thread when connecting etc.

rmn30 committed:
HughV2 cheriot: add support for cheriot.demo local server.

Add trust anchor for a self-signed cert used with local server and
modify ifdefing to allow selecting between three different servers at
compile time.  Print a message showing the server being used to assist
with debugging.

rmn30 committed:
Add set_brightness on SonataLCD for controlling backlight.

rmn30 committed:
Add include required for heap_available

rmn30 committed:
HughV2: Enable controlling LCD brightness with joystick.

I hypothesise that this might help with scanning on the demo tablet.

rmn30 committed:
HughV2: tweaks to mqtt reconnection logic.

Two tweaks to improve stability:

1) Rate limit reconnection attempts to one per second. This prevents hammering the server in case the connect fails immediately.

2) Don't use an unlimited timeout. This ensures we eventually time out in case we something goes wrong.

rmn30 committed:
HughV2: update letsencrypt certificate to use ISRG Root X1 cert.

This seems to be required to connect to demo.cheriot.org now.

rmn30 committed:
Bump submodules.

nwf commented on HughTheLightbulbV2/HughController/app/src/main/java/org/cheriot/hughcontroller/MainActivity.java:
```diff
@@ -102,15 +113,27 @@ private void initialiseMQTT()
+            String[] serverParts = selectedServer.split(":");
+            String server = serverParts[0];
+            int port = Integer.valueOf(serverParts[1]);
+            boolean use_ssl = port != 1883;
```
Should we be using URIs for this?  Is that something the HiveMQ client can just do?

nwf commented on HughTheLightbulbV2/HughController/app/src/main/java/org/cheriot/hughcontroller/MainActivity.java:
```diff
@@ -131,18 +154,31 @@ protected void onCreate(Bundle savedInstanceState) {
         });
 
-        initialiseMQTT();
+//        initialiseMQTT();
```
No?  Why not?

rmn30 replied:
Looks like I left in a few other versions I tried. We're not really doing this the right way. We're using the synchronous API on the main thread which causes the app to hang if anything goes wrong connecting. 

I think we should make a screen that launches when the app starts to select a server and then we should either use the asynchronous API or I should figure out how to spawn a background task to do the connecting.

nwf reviewed (approved):
Seems OK by me, just some Java-side questions.

rmn30 committed:
Include allocator.h where required.

We used to pick this up by including stdlib.h but no longer.

rmn30 committed:
Fix clang warning about static in explicit template instatiation.

rmn30 committed:
Replace deprecated SKey with TokenKey.

rmn30 committed:
Fix config capability macros.

Clang no longer accepts anonymous structs in this location.

rmn30 committed:
Add an explicit allocator capability argument to ConfigConsuer::run.

Libraries no longer have MALLOC_CAPABILITY defined by default so we
require that the calling compartment pass us one.
